streams/shared/src/main/scala/zio/stream/Structure.scala

```diff
@@ -0,0 +1,148 @@
+
+      for {
+        reservation <- proc.reserve
+        _           <- currFin.set(reservation.release)
+        _           <- restore(reservation.acquire.tap(currPull.set))
+      } yield ()
+    }
+
+  private[this] def pull(
+    currStruct: Ref[Structure[R, E, Any]],
+    currFin: Ref[Exit[_, _] => URIO[R, _]],
+    currPull: Ref[Pull[R, E, Any]],
+    stack: Ref[List[(Pull[R, E, _], Exit[_, _] => URIO[R, _], Structure[R, E, _])]]
+  ): Pull[R, E, A] =
+    currPull.get.flatten.foldM(
+      {
+        case e @ Some(_) => ZIO.fail(e)
+        case None =>
+          ZIO.uninterruptibleMask {
+            restore =>
+              currStruct.get.flatMap {
+                case Concat(_, tl) =>
+                  val tl0 = tl()
+                  currFin.get.flatMap(_.apply(Exit.interrupt)) *>
+                    currStruct.set(tl0) *> restore(open(tl0, currFin, currPull).mapError(Some(_)))
+                case _ =>
+                  stack.get.flatMap {
+                    case Nil => Pull.end
+                    case (p, fin, s) :: rest =>
+                      currFin.get.flatMap(_.apply(Exit.interrupt)) *>
+                        stack.set(rest) *>
+                        currPull.set(p) *>
+                        currFin.set(fin) *>
+                        currStruct.set(s)
+                  }
+              }
+          } *> pull(currStruct, currFin, currPull, stack)
+      },
+      v =>
+        currStruct.get.flatMap {
+          case Iterator(_) | Concat(_, _) =>
+            UIO.succeed(v.asInstanceOf[A])
+
+          case Effect(_) =>
+            currPull.set(Pull.end).as(v.asInstanceOf[A])
+
+          case FlatMap(Effect(_), k) =>
+            // This combination is special cased because with a stream represented
+            // by Effect, we don't have to push anything onto the stack when switching
+            // to the stream produced by k for the resulting element. This is because
+            // Effect represents a stream with one element that has no finalizer, so
+            // there's no reason for us to re-visit it after evaluation.
+            ZIO.uninterruptibleMask { restore =>
+              val next = k.asInstanceOf[Any => Structure[R, E, Any]](v)
+
+              currStruct.set(next) *>
+                restore(open(next, currFin, currPull).mapError(Some(_)))
+            } *> pull(currStruct, currFin, currPull, stack)
+
+          case FlatMap(_, k) =>
+            ZIO.uninterruptibleMask { restore =>
+              val next = k.asInstanceOf[Any => Structure[R, E, Any]](v)
+
+              (for {
+                st  <- stack.get
+                p   <- currPull.get
+                fin <- currFin.get
+                s   <- currStruct.get
+                _   <- stack.set((p, fin, s) :: st)
+                _   <- currStruct.set(next)
+              } yield ()) *> restore(open(next, currFin, currPull).mapError(Some(_)))
+            } *> pull(currStruct, currFin, currPull, stack)
+        }
```

Comment:
This frame is going to be on the hot path? I wonder if using a `tag` akin to `zio.tag` would be a good optimisation here.